tensorflow/core/kernels/data/concatenate_dataset_op_test.cc

```diff
@@ -0,0 +1,378 @@
+  int64 expected_cardinality;
+  std::vector<int> breakpoints;
+} TestCases[] = {
+    // Test case 1: same shape.
+    {/*input_tensors*/ {{DatasetOpsTestBase::CreateTensor<int64>(
+                             TensorShape{2, 2}, {1, 2, 3, 4}),
+                         DatasetOpsTestBase::CreateTensor<int64>(
+                             TensorShape{2, 2}, {5, 6, 7, 8})},
+                        {DatasetOpsTestBase::CreateTensor<int64>(
+                             TensorShape{2, 2}, {11, 12, 13, 14}),
+                         DatasetOpsTestBase::CreateTensor<int64>(
+                             TensorShape{2, 2}, {15, 16, 17, 18})}},
+     /*expected_outputs*/
+     {DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {1, 2}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {5, 6}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {3, 4}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {7, 8}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {11, 12}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {15, 16}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {13, 14}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {17, 18})},
+     /*expected_output_dtypes*/ {DT_INT64, DT_INT64},
+     /*expected_output_shapes*/
+     {PartialTensorShape({2}), PartialTensorShape({2})},
+     /*expected_cardinality*/ 4,
+     /*breakpoints*/ {0, 2, 5}},
+    // Test case 2: different shape.
+    {/*input_tensors*/ {{DatasetOpsTestBase::CreateTensor<int64>(
+                             TensorShape{2, 3}, {1, 2, 3, 4, 5, 6}),
+                         DatasetOpsTestBase::CreateTensor<int64>(
+                             TensorShape{2, 2}, {7, 8, 9, 10})},
+                        {DatasetOpsTestBase::CreateTensor<int64>(
+                             TensorShape{2, 2}, {11, 12, 13, 14}),
+                         DatasetOpsTestBase::CreateTensor<int64>(
+                             TensorShape{2, 1}, {15, 16})}},
+     /*expected_outputs*/
+     {DatasetOpsTestBase::CreateTensor<int64>(TensorShape{3}, {1, 2, 3}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {7, 8}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{3}, {4, 5, 6}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {9, 10}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {11, 12}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{1}, {15}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {13, 14}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{1}, {16})},
+     /*expected_output_dtypes*/ {DT_INT64, DT_INT64},
+     /*expected_output_shapes*/
+     {PartialTensorShape({-1}), PartialTensorShape({-1})},
+     /*expected_cardinality*/ 4,
+     /*breakpoints*/ {0, 2, 5}}};
+
+struct ConcatenateDatasetOpTestHelper : ConcatenateDatasetOpTest {
```

Comment:
"Use a `struct` only for passive objects that carry data; everything else is a `class`." https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes

Likewise for most of your other uses of `struct`